### ISSUE TITLE
'coincident' combinator

### ISSUE THREAD

**Issue by mitchellwrosen** [HeinrichApfelmus/reactive-banana#158]
I've occasionally found this combinator useful:

```haskell
-- | Only emit event occurrences coincident with the second event.
coincident :: Event a -> Event b -> Event a
coincident e1 e2 = 
  filterRight                  -- (3) Filter down to only coincident events
    (unionWith
      (\(Left x) _ -> Right x) -- (2) Re-tag coincident Lefts as Rights
      (Left <$> e1)            -- (1) Left means "not coincident"
      (unsafeCoerce e2))
 where
  filterRight e = filterJust (either (const Nothing) Just <$> e)
```

I wonder if anyone else has, too?

**Comment by nickkuk** [comment]
What about such simplification?
```haskell
coincident :: Event a -> Event b -> Event a
coincident e1 e2 = filterJust (unionWith const (Nothing <$ e2) (Just <$> e1))
```

**Comment by mitchellwrosen** [comment]
That's the XOR of mine :)

**Comment by nickkuk** [comment]
Ah, sorry, I misunderstood :smile:

**Comment by nickkuk** [comment]
One could suggest more general function
```haskell
intersectionWith :: (a -> b -> c) -> Event a -> Event b -> Event c
```

**Comment by mitchellwrosen** [comment]
Yup, and the implementation is just a variation of `unionWithP`:

```haskell
intersectionWithP :: (a -> b -> c) -> Pulse a -> Pulse b -> Build (Pulse c)
intersectionWithP f px py = do
    p <- newPulse "intersectionWithP" $ eval <$> readPulseP px <*> readPulseP py
    p `dependOn` px
    p `dependOn` py
    return p
  where
    eval (Just x) (Just y) = Just (f x y)
    eval _ _ = Nothing
```

though the name is kind of long... I wish `unionWith` was called `union` instead!

**Comment by nickkuk** [comment]
I think these names are good, because they correspond to similar functions from e.g. Data.Map.

**Comment by mitchellwrosen** [comment]
Well there used to be a `union` function before version 1.0, so the name `unionWith` made more sense. Now that `union` no longer exists, I'd prefer to simply rename `unionWith` to save 4 characters... but that's just a personal preference, and anyways, beside the point of this ticket :)

**Comment by nickkuk** [comment]
@mitchellwrosen I see, that you added PR about intersectionWith. What do you think about even more general function [mergeWithKey](https://hackage.haskell.org/package/containers-0.5.10.2/docs/Data-IntMap-Lazy.html#v:mergeWithKey)?
In the case of reactive-banana its type will look like
```[haskell]
mergeWith :: (a -> b -> Maybe c) -> (Event a -> Event c) -> (Event b -> Event c) -> Event a -> Event b -> Event c
```

**Comment by mitchellwrosen** [comment]
I think its `reactive-banana` type would actually be this:

```haskell
mergeWith 
  :: (a -> b -> Maybe c) 
  -> (a -> Maybe c) 
  -> (b -> Maybe c) 
  -> Event a 
  -> Event b 
  -> Event c
```

Looks very general, I'm not sure how useful it would be in practice. Still, `unionWith` and `intersectionWith` could both be implemented in terms of it:

```haskell
unionWith :: (a -> a -> a) -> Event a -> Event a -> Event a
unionWith f e1 e2 = mergeWith (\x y -> Just (f x y)) Just Just

intersectionWith :: (a -> b -> c) -> Event a -> Event b -> Event c
intersectionWith f e1 e2 = mergeWith (\x y -> Just (f x y)) (const Nothing) (const Nothing)
```

Hmm :thinking:

**Comment by nickkuk** [comment]
Aha! I start to write the message that "two inner `-> Event c` are superfluous" :smile:

**Comment by nickkuk** [comment]
Also my broken [coincident](https://github.com/HeinrichApfelmus/reactive-banana/issues/158#issuecomment-330062603) (that is acutally [difference](https://hackage.haskell.org/package/containers-0.5.10.2/docs/Data-IntMap-Lazy.html#v:difference) and can be generalized to [differenceWith](https://hackage.haskell.org/package/containers-0.5.10.2/docs/Data-IntMap-Lazy.html#v:differenceWith)) can be implemented in terms of `mergeWith`.

**Comment by nickkuk** [comment]
@mitchellwrosen, let me point out the situation when one needs [mergeWithKey](https://hackage.haskell.org/package/containers-0.5.10.2/docs/Data-IntMap-Lazy.html#v:mergeWithKey) and when [intersectionWith](https://hackage.haskell.org/package/containers-0.5.10.2/docs/Data-IntMap-Lazy.html#v:intersectionWith), [unionWith](https://hackage.haskell.org/package/containers-0.5.10.2/docs/Data-IntMap-Lazy.html#v:unionWith), [differenceWith](https://hackage.haskell.org/package/containers-0.5.10.2/docs/Data-IntMap-Lazy.html#v:differenceWith) are not enough.

If you want to work with sparse vectors over the field of rational numbers, you can just filter it by non zero elements and store in `IntMap Rational`. So `newtype Vector = Vector (IntMap Rational)`. But to implement addition or subtraction of such vectors it is convenient and more performant to use `mergeWithKey`, because you get `unionWith` + filtering of non zeros in one function. In terms of your function `mergeWith` it will looks like
```[haskell]
add = mergeWith (\a b -> filterNonZero (a + b)) Just Just
sub = mergeWith (\a b -> filterNonZero (a - b)) Just (Just . negate)
filterNonZero a = if a == 0 then Nothing else Just a
```

I believe that it is not hard to find use cases for `mergeWith` in reactive programming.

**Comment by ocharles** [comment]
@mitchellwrosen Now that we have `merge`, I believe we have

```haskell
coincident :: Event a -> Event b -> Event a
coincident e1 e2 = filterJust $ merge e1 e2 <&> \case
  These a _ -> Just a
  _         -> Nothing
```

Do you still think we should add this to `R.B.Combinators`?

**Comment by mitchellwrosen** [comment]
Hmm... I'm not sure :)

On one hand, I do find I need this combinator, or something like it, a lot.

But on the other hand, I don't like how the order of arguments isn't clear to a reader.

**Comment by HeinrichApfelmus** [comment]
> But on the other hand, I don't like how the order of arguments isn't clear to a reader.

How about using an adjective, e.g. a present participle?

```
intersecting :: Event a -> Event b -> Event a

… (ea `intersecting` eb) …
```
